File: java/ValidSudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        // check rows
        for (int i = 0; i < 9; i++){
            boolean[] chs = new boolean[9];
            for (int j = 0; j < 9; j++){
                if (board[i][j] == '.'){
                    continue;
                }
                else if (chs[board[i][j] - '1']){
                    return false;
                }
                else{
                    chs[board[i][j] - '1'] = true;
                }
            }
        }
        // check columns
        for (int j = 0; j < 9; j++){
            boolean[] chs = new boolean[9];
            for (int i = 0; i < 9; i++){
                if (board[i][j] == '.'){
                    continue;
                }
                else if (chs[board[i][j] - '1']){
                    return false;
                }
                else{
                    chs[board[i][j] - '1'] = true;
                }
            }
        }
        // check sub-boxes
        for (int m = 0; m < 3; m++){
            for (int n = 0; n < 3; n++){
                boolean[] chs = new boolean[9];
                for (int i = 0; i < 3; i++){
                    for (int j = 0; j < 3; j++){
                        if (board[3 * m + i][3 * n + j] == '.'){
                            continue;
                        }
                        if (chs[board[3 * m + i][3 * n + j] - '1']){
                            return false;
                        }
                        else{
                            chs[board[3 * m + i][3 * n + j] - '1'] = true;
                        }
                    }
                }
            }
        }
        
        return true;
    }
}

File: python/ValidSudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        # check rows
        return self.isRowValid(board) and self.isColumnValid(board) and self.isSquareValid(board)
    
    def isRowValid(self, board):
        for row in board:
            if not self.isUnitValid(row):
                return False
        return True
        
    def isColumnValid(self, board):
        for col in zip(*board):
            if not self.isUnitValid(col):
                return False
        return True
        
    def isSquareValid(self, board):
        for i in (0, 3, 6):
            for j in (0, 3, 6):
                square = [board[x][y] for x in range(i, i + 3) for y in range(j, j + 3)]
                if not self.isUnitValid(square):
                    return False
        return True
    
    def isUnitValid(self, unit):
        res = [i for i in unit if i != '.']
        return len(res) == len(set(res))
